src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -621,10 +709,21 @@ protected override void Dispose(bool disposing)
                 public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken)
                 {
                     Http2Stream http2Stream = _http2Stream;
-                    Task t = http2Stream != null ?
-                        http2Stream.SendDataAsync(buffer, cancellationToken) :
-                        Task.FromException(new ObjectDisposedException(nameof(Http2WriteStream)));
-                    return new ValueTask(t);
+
+                    if (http2Stream == null)
+                    {
+                        return new ValueTask(Task.FromException(new ObjectDisposedException(nameof(Http2WriteStream))));
+                    }
+
+                    // We got rejected while sending request.
+                    if (http2Stream._response != null && (int)http2Stream._response.StatusCode >= 300)
```

Comment:
This is really just a special case of #35652, right?

Seems like what we need here is a general approach to handling exceptions when sending the request body. I.e. not specific to 100-continue.

Reply:
I think we need to do a couple things, to address both this specific abort case and the general case of #35652:

(1) Just throw ObjectDisposedException here. We don't need to detect the specific case of 100-continue failure.
(2) Catch exceptions during the send of the request, and send a RST_STREAM at that point. Also, probably set the state to Aborted when this happens.

I think if we do those two things, we'll address all the issues here.

Reply:
Also, we need to remove the stream from the stream dictionary when this happens. Not sure if we do this today -- I think we don't.

Reply:
I fixed SendAsync() to dispose AND remove stream just like SendHeadersAsync() does.

Reply:
as far as the 100COntinue: 
What I wanted to do is abandon sending request if we get reject but keep reading response and provide response with code and message to caller rather than throwing IO (or other) exception.

I could move SendRstStreamAsync() to SendAsync(). But I think it would be good to track if we decided to stop sending or the ObjectDisposed is result of protocol or IO error if you agree with behavior described above.

Reply:
Yeah, I see your point -- this is not really the same case as other abort cases, because we don't want to fail the request here. Am I understanding you correctly?

Let me make sure I understand the case we're dealing with here...

We already will abort sending the response if we get a >= 300 response before the timer expires, right? We have logic above in ReadResponseHeadersAsync to handle this.

So what case is this addressing? Is it the case where we started sending because the timer expired, and then later received a >= 300 response? For HTTP/1.1, we just finish sending the request body in this case. It seems reasonable to do the same here, for HTTP2. I guess you could argue that it would be better to silently abort the request body in this case, but I'm not sure it matters much. (We can't abort the request body for HTTP/1.1 without killing the connection, so we don't ever do this currently.)

I guess this raises the question of why this stream is disposed at this point? If it wasn't disposed, we'd just keep sending the body in this case and no error would happen.

Note this issue isn't really specific to 100-continue; it's always possible to receive an "early" response, i.e. before we finish sending the request. If we're going to try to abort sending the request in these cases, we should probably do it for non 100-continue cases as well. It's unclear to me if we actually are doing this or not here...

Reply:
By the way, I don't see where we are sending RST_STREAM in the case where we get a >= 300 response *before* the timer fires. See my comment in SendRequestBodyWithExpect100ContinueAsync above.

Reply:
> Yeah, I see your point -- this is not really the same case as other abort cases, because we don't want to fail the request here. Am I understanding you correctly?
> 
yes. 


> Let me make sure I understand the case we're dealing with here...
> 
> We already will abort sending the response if we get a >= 300 response before the timer expires, right? We have logic above in ReadResponseHeadersAsync to handle this.
> 
yes. We will not start sending request body if we get negative response before timer expires. 


> So what case is this addressing? Is it the case where we started sending because the timer expired, and then later received a >= 300 response? For HTTP/1.1, we just finish sending the request body in this case. It seems reasonable to do the same here, for HTTP2. I guess you could argue that it would be better to silently abort the request body in this case, but I'm not sure it matters much. (We can't abort the request body for HTTP/1.1 without killing the connection, so we don't ever do this currently.)

yes, we can do that. However the way I read #31312 was that you want us to send abort and stop sending. If we do no stop, wee do not need to send RST, right?

> 
> I guess this raises the question of why this stream is disposed at this point? If it wasn't disposed, we'd just keep sending the body in this case and no error would happen.
> 
> Note this issue isn't really specific to 100-continue; it's always possible to receive an "early" response, i.e. before we finish sending the request. If we're going to try to abort sending the request in these cases, we should probably do it for non 100-continue cases as well. It's unclear to me if we actually are doing this or not here...

One option I was thinking about is simply setting bool flag so we do not need to guess based on exception. Would that be more clean and reasonable?

Reply:
> However the way I read #31312 was that you want us to send abort and stop sending. If we do no stop, wee do not need to send RST, right?

Yeah, my comment (2) in the linked issue is about doing exactly this. I don't think it's strictly necessary. It's kind of a corner case: the timer expired so we started sending, but then we received a >= 300 status code. Doesn't seem that common in practice.

So I think we could certainly live without it. But if we're going to add the handling for this, then we may as well handle the same case for non 100-continue as well, i.e. we receive a >= 300 status code while sending the request body.

> One option I was thinking about is simply setting bool flag so we do not need to guess based on exception. Would that be more clean and reasonable?

Yeah, something like this seems reasonable to me. That is, if we receive a final status code while we are in the middle of sending the request body, then:
(1) Set a flag like "_abortRequestBody = true"
(2) Dispose the write stream so it will cause an ObjectDisposedException on the next write
(3) Catch the exception during request body send, and if _abortRequestBody is true, just eat it
(4) We also need to send RST_STREAM when we catch the exception, but we need to do that regardless of whether we are eating the exception or not. This will contribute to #35652, though there may be additional work we need to do to get that working.